src/main/java/org/assertj/core/internal/Maps.java

```diff
@@ -813,38 +816,72 @@ public void assertHasSameSizeAs(AssertionInfo info, Map<?, ?> map, Map<?, ?> oth
   private <K, V> void compareActualMapAndExpectedKeys(Map<K, V> actual, K[] keys, Set<K> notExpected,
                                                       Set<K> notFound) {
 
-    Map<K, V> actualEntries = new LinkedHashMap<>(actual);
+    // We can't use a copy of actual since we might change the implementation of the map
+    Set<K> foundKeys = new LinkedHashSet<>();
     for (K key : keys) {
-      if (actualEntries.containsKey(key)) {
+      if (actual.containsKey(key)) {
         // this is an expected key
-        actualEntries.remove(key);
+        foundKeys.add(key);
       } else {
         // this is a not found key
         notFound.add(key);
       }
     }
-    // All remaining keys from actual copy are not expected entries.
-    notExpected.addAll(actualEntries.keySet());
+
+    // We now iterate through the actual entries in order to detect the unexpected ones
+    for (Map.Entry<K, V> entry : actual.entrySet()) {
+      boolean keyNotExpected = true;
+      Iterator<K> foundKeysIterator = foundKeys.iterator();
+      while (foundKeysIterator.hasNext()) {
+        K foundKey = foundKeysIterator.next();
+        V foundValue = actual.get(foundKey);
+        if (areEqual(foundValue, entry.getValue())) {
```

Comment:
I'm not sure to see why we need to compare values when the assertion is only checking keys, am I missing something here?

Reply:
I think we have the same issue as with `containsExactly` since we have to compare keys and we can't access actual's map key comparison strategy.

This test fails as it should but it does not report the correct error:
```java
    actual = new CaseInsensitiveMap<>();
    actual.put("NAME", "green");
    actual.put("cool", "green");
    // THEN
    maps.assertContainsOnlyKeys(someInfo(), actual, "Name", "Color");
```
error is:
```
Expecting actual:
  {"cool"="green", "name"="green"}
to contain only following keys:
  ["Name", "Color"]
keys not found:
  ["Color"]
and keys not expected:
  ["name"]
```
"keys not found" is correctly reported but "keys not expected" should have shown "cool", it reports "name" instead which actually is a map key.

Reply:
I guess the only way forward is to find a way to clone actual's map which might be possible using introspection to call a a no arg constructor and then populate the cloned map. Question is now: is there always a no arg constructor ...

Reply:
You are right. What we can also do is to support sime well known case insensitive maps behind the scenes, and for everything else create a normal copy. Reusing constructors might lead to using unmodifiable / immutable maps. I can give it another go on Sunday afternoon, no access to a laptop before that

Reply:
> Question is now: is there always a no arg constructor ...

[Objenesis](http://objenesis.org/) could deal with it and I think it would be fine to have it as a shaded dependency

Reply:
I was thinking we could try to instantiate the map (I'm fine using Objenesis) and if it fails for whatever reason use a simple `LinkedHashMap` as in the existing implementation, that won't completely solve the issue but will work for a vast majority of use cases, finally we would mention this limitation in the javadoc.

Reply:
I have tried Objenesis and that did not go well with `CaseInsensitiveMap`, creating the instance worked but would not initialize `transient HashEntry<K, V>[] data;` which is accessed  later on when putting actual entries in the new map.

Interestingly it works with `ConstructorInvoker` (some old internal class in our codebase) and it worked.

Here's the stack trace for the record:
```
java.lang.NullPointerException
	at org.apache.commons.collections4.map.AbstractHashedMap.ensureCapacity(AbstractHashedMap.java:627)
	at org.apache.commons.collections4.map.AbstractHashedMap._putAll(AbstractHashedMap.java:325)
	at org.apache.commons.collections4.map.AbstractHashedMap.putAll(AbstractHashedMap.java:304)
	at org.assertj.core.internal.Maps.compareActualMapAndExpectedKeys(Maps.java:823)
```